### PR TITLE
Preliminary Prototype of Searchable Encryption

### PR DESCRIPTION
**Pull request by jxguan**
Implements a basic searchable encryption scheme that runs locally with clients and a server. Simulates the network by including latency and bandwidth in the logging.

### CODE REVIEW

**Comment by jxguan** [comment]
For code review, probably this order would be the easiest:
util -> index -> indexer -> searcher -> logger -> server -> client -> main.go

File: README.md [hunk 1]
```diff
@@ -1,4 +1,39 @@
...
+go run main.go
+```
+
+Or better, with time logging:
+
+```
+go run main.go --enable_logger
+```
+
+The clients have a defualt mount point of `.client_fs/` and the server has a
```
Comment (strib) [inline README.md]: "defualt" -> "default"

Comment (strib) [inline README.md]: Also, "mount point" means something specific in reference to a file system.  I think here you just mean a directory for state storage?

File: client/client.go [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+	"strconv"
+	"strings"
+)
+
+// Client stores the necessary information for a client.
+type Client struct {
+	mountPoint    string                      // Mount point for the client where all the files are stored
+	server        *server.Server              // The server that this client is connected to
+	indexer       *indexer.SecureIndexBuilder // The indexer for the client
+	lookupTable   map[string]string           // A map from docuemnt ids to actual filenames
```
Comment (strib) [inline client/client.go]: "docuemnt" -> "document"

File: util/util.go [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+func GenerateSalts(numKeys, lenSalt int) (salts [][]byte) {
+	if lenSalt < 8 {
+		fmt.Println("Error in generating the salts: lenSalt must be at leat 8")
+	}
+	salts = make([][]byte, numKeys)
+	for i := 0; i < numKeys; i++ {
+		salts[i] = make([]byte, lenSalt)
+		_, err := rand.Read(salts[i])
+		if err != nil {
+			fmt.Println("Error in generating the salts: ", err)
```
Comment (strib) [inline util/util.go]: It's better practice to always return the error, and let the highest level possible deal with it.  This function can't really be re-used as a library function since it just prints out the errors.

File: index/secure_index.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
+package index
+
+import (
+	"crypto/sha256"
+	"crypto/sha512"
+	"encoding/binary"
+	"fmt"
+	"github.com/jxguan/go-datastructures/bitarray"
```
Comment (strib) [inline index/secure_index.go]: non-standard-lib packages should go in a separate block of imports, after a blank line. Like:

``` go
import (
    "crypto/sha256"
    ...
    "hash"

    "github.com/jxguan/go-datastructures/bitarray"
)
```

File: index/secure_index.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+// SecureIndex defines the elements in a secure index.
+type SecureIndex struct {
+	BloomFilter bitarray.BitArray // The blinded bloom filter, which is the main part of the index.
+	DocID       int               // The document ID that this index is for.
+	Size        uint64            // The number of buckets in the bloom filter.
+	Hash        func() hash.Hash  // The hash function to be used for HMAC.
+}
+
+// Marshal serializes a SecureIndex into a byte slice.
+func (si *SecureIndex) Marshal() []byte {
```
Comment (strib) [inline index/secure_index.go]: Implementing the [`encoding.BinaryMarshaler`](https://golang.org/pkg/encoding/#BinaryMarshaler) interface would be more standard.

(Same with Unmarshal below.)

File: index/secure_index.go [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+	bfBytes, err := bitarray.Marshal(si.BloomFilter)
```
Comment (strib) [inline index/secure_index.go:23]: Same here (I assume this is a custom method you added to your fork).

File: indexer/secure_index_builder.go [hunk 7]
```diff
@@ -0,0 +1,94 @@
+package indexer
+
+import (
+	"bufio"
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/binary"
+	"github.com/jxguan/go-datastructures/bitarray"
```
Comment (strib) [inline indexer/secure_index_builder.go]: Same comment here about github/golang package imports going in their own block.

File: indexer/secure_index_builder.go [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+// indexes for the documents.
+type SecureIndexBuilder struct {
+	numKeys      int                   // The number of keys.  This is also the number of PRFs.
+	keys         [][]byte              // The keys for the PRFs. Derived from the masterSecret and the salts.
+	hash         func() hash.Hash      // The hash function to be used for HMAC.
+	trapdoorFunc func(string) [][]byte // The trapdoor function for the words
+	size         uint64                // The size of each index, i.e. the number of buckets in the bloom filter.  Smaller size will lead to higher false positive rates.
+}
+
+// CreateSecureIndexBuilder instantiates a `secureIndexBuilder`.  Sets up the
```
Comment (strib) [inline indexer/secure_index_builder.go]: "secureIndexBuilder" -> "SecureIndexBuilder"

File: indexer/secure_index_builder.go [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+			mac.Write([]byte(string(docID)))
+			// Ignore the error as we need to truncate the 256-bit hash into 64 bits
+			codeword, _ := binary.Uvarint(mac.Sum(nil))
+			bf.SetBit(codeword % sib.size)
+		}
+	}
+	return bf, len(words)
+}
+
+// Blinds the bloom filter by setting randon bits to be on for `numIterations`
```
Comment (strib) [inline indexer/secure_index_builder.go]: "randon" -> "random"

File: index/secure_index.go (hunk 6 above)
Comment (jxguan) [inline index/secure_index.go:23]: Oh it's actually created by the people who wrote the package. I forked it to fix a bug in GetCapacity.

File: indexer/secure_index_builder.go [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+	bf := bitarray.NewSparseBitArray()
+	words := make(map[string]bool)
+	for scanner.Scan() {
+		word := scanner.Text()
+		words[word] = true
+		trapdoors := sib.trapdoorFunc(word)
+		for _, trapdoor := range trapdoors {
+			mac := hmac.New(sib.hash, trapdoor)
+			mac.Write([]byte(string(docID)))
+			// Ignore the error as we need to truncate the 256-bit hash into 64 bits
```
Comment (strib) [inline indexer/secure_index_builder.go]: Maybe just check that the error matches what you expect?  Or you could use bit-shifting for truncation?

File: indexer/secure_index_builder.go [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+// hash function, and derives the keys from the master secret and salts by using
+// PBKDF2.  Finally, sets up the trapdoor function for the words.
+func CreateSecureIndexBuilder(h func() hash.Hash, masterSecret []byte, salts [][]byte, size uint64) *SecureIndexBuilder {
+	sib := new(SecureIndexBuilder)
+	sib.keys = make([][]byte, len(salts))
+	for index, salt := range salts {
+		sib.keys[index] = pbkdf2.Key(masterSecret, salt, 4096, 32, sha256.New)
+	}
+	sib.hash = h
+	sib.numKeys = len(salts)
```
Comment (strib) [inline indexer/secure_index_builder.go]: Do we need to track this?  Can't we just use `len(sib.keys)`?

File: logger/logger.go [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+// supposed to be done.  Logging is disabled by default.  To turn them on, call
+// `logger.Enable()`.  A typical use case would be:
+//
+// func test() {
+//   logger.Enable()
+//   logger.Start("test")
+//
+//   ... Some Work ...
+//
+//   logger.AddTime(time.Minute * 2)  // Simultaes two minutes doing some work
```
Comment (strib) [inline logger/logger.go]: "Simultaes" -> "Simulates"

File: logger/logger.go [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+//
+//   logger.Log("test")
+// }
+
+// timeLogs stores the correspondance between entry names and the start time of
+// each entry.  The start time would be adjusted accordingly to account for the
+// `Addtime` function.
+var timeLogs = make(map[string]time.Time)
+
+// enabled determines whether the logger is enables.  Defaulted to false.
```
Comment (strib) [inline logger/logger.go]: "enables" -> "enabled"

File: logger/logger.go [hunk 14]
```diff
@@ -0,0 +1,77 @@
...
+//   logger.AddTime(time.Minute * 2)  // Simultaes two minutes doing some work
+//
+//   ... Other Work ...
+//
+//   logger.Log("test")
+// }
+
+// timeLogs stores the correspondance between entry names and the start time of
+// each entry.  The start time would be adjusted accordingly to account for the
+// `Addtime` function.
```
Comment (strib) [inline logger/logger.go]: "Addtime" -> "AddTime"

File: server/server.go [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+	s.salts = util.GenerateSalts(r, lenSalt)
+	s.numFiles = 0
+	s.mountPoint = mountPoint
+	s.writeToFile()
+	return s
+}
+
+// CreateServerWithLog behaves the same as `CreateServer`, except for that it
+// also sets the logging parameters for the server.
+func CreateServerWithLog(numClients, lenMS, lenSalt int, mountPoint string, fpRate float64, numUniqWords uint64, latency time.Duration, bandwidth int) *Server {
```
Comment (strib) [inline server/server.go]: It's a little weird to call these "logging parameters" (and to suffix the function with "WithLog"), since there are really network simulation parameters, which are conceptually different from logging.  Maybe "WithNetConfig" instead?

File: server/server.go [hunk 16]
```diff
@@ -0,0 +1,229 @@
...
+
+	file.Close()
+}
+
+// AddFile adds a file with `content` to the server with the document ID equal
+// to the number of files currently in the server and updates the count.
+// Returns the document ID.
+func (s *Server) AddFile(content []byte) int {
+	logger.AddTime(s.latency * 2)
+	logger.AddTime(time.Millisecond * time.Duration(float64(len(content))*1.5*8*1000/float64(s.bandwidth)))
```
Comment (strib) [inline server/server.go:120]: Can you document why "1.5" is needed here (to account for encryption, I assume).

Could you use `time.Second` instead of `time.Millisecond`, and then drop the `*1000`?  Or maybe you need to do it this way because otherwise rounding causes problems?

File: server/server.go [hunk 17]
```diff
@@ -0,0 +1,229 @@
...
+	s.writeToFile()
+	return s.numFiles - 1
+}
+
+// GetFile returns the content of the document with `docID`.  Behavior is
+// undefined if the docID is invalid (out of range).
+func (s *Server) GetFile(docID int) []byte {
+	logger.AddTime(s.latency * 2)
+	content, _ := ioutil.ReadFile(path.Join(s.mountPoint, strconv.Itoa(docID)))
+	logger.AddTime(time.Millisecond * time.Duration(float64(len(content))*1.5*8*1000/float64(s.bandwidth)))
```
Comment (strib) [inline server/server.go:134]: Looks like you could use a helper function to get the extra time to add given a byte array.

File: server/server.go [hunk 18]
```diff
@@ -0,0 +1,229 @@
...
+}
+
+// readIndex loads an index from the disk.
+func (s *Server) readIndex(docID int) index.SecureIndex {
+	input, _ := ioutil.ReadFile(path.Join(s.mountPoint, strconv.Itoa(docID)+".index"))
+	si := index.Unmarshal(input)
+	return si
+}
+
+// SearchWord searchers the server for a word with `trapdoors`.  Returns a list
```
Comment (strib) [inline server/server.go]: "searchers" -> "searches"

File: server/server_test.go [hunk 19]
```diff
@@ -0,0 +1,228 @@
...
+
+// TestSearchWord tests the `SearchWord` function.  Checks that the correct set
+// of files are returned when searching for a word on the server.
+func TestSearchWord(t *testing.T) {
+	s, dir := createTestServer(5, 8, 8, 0.000001, uint64(100000))
+	defer os.RemoveAll(dir)
+	sib := indexer.CreateSecureIndexBuilder(sha256.New, calculateMasterSecret(0, s.keyHalves[0]), s.salts, s.size)
+
+	files := []string{
+		"charmander pikachu bulbasaur",
```
Comment (strib) [inline server/server_test.go:172]: I thought "zubat" was supposed to be the most common pokemon.

File: README.md (hunk 1 above)
Comment (jxguan) [inline README.md]: Fixed

File: client/client.go (hunk 2 above)
Comment (jxguan) [inline client/client.go]: Fixed

File: client/client.go [hunk 20]
```diff
@@ -0,0 +1,142 @@
...
+	h.Write([]byte(strconv.Itoa(clientNum)))
+	serverKeyHalf := s.GetKeyHalf(clientNum)
+	ms := util.XorBytes(h.Sum(nil), serverKeyHalf, len(serverKeyHalf))
+	c.indexer = indexer.CreateSecureIndexBuilder(sha256.New, ms, s.GetSalts(), s.GetSize())
+
+	// Initializes the lookup table
+	// NOTE: Factor out and add decryption
+	c.lookupTable = make(map[string]string)
+	c.reverseLookup = make(map[string]string)
+	if tableContent, found := s.ReadLookupTable(); found {
```
Comment (strib) [inline client/client.go:50]: As we talked about today, I think it would be tough to synchronize the lookup table to all the clients this way.  We're better off using the full file name as the document ID, and updating that on the server whenever it changes.

File: util/util.go (hunk 3 above)
Comment (jxguan) [inline util/util.go]: Fixed

File: index/secure_index.go (hunk 4 above)
Comment (jxguan) [inline index/secure_index.go]: Fixed

File: indexer/secure_index_builder.go (hunk 7 above)
Comment (jxguan) [inline indexer/secure_index_builder.go]: Fixed

File: client/client.go (hunk 20 above)
Comment (strib) [inline client/client.go:50]: (This can be a separate PR, just noting it here.)

File: indexer/secure_index_builder.go (hunk 8 above)
Comment (jxguan) [inline indexer/secure_index_builder.go]: Fixed

File: indexer/secure_index_builder.go (hunk 9 above)
Comment (jxguan) [inline indexer/secure_index_builder.go]: Fixed

File: indexer/secure_index_builder.go (hunk 10 above)
Comment (jxguan) [inline indexer/secure_index_builder.go]: Hmm it's actually not even an error, just the number of bytes read. I should probably just remove the comment, which is confusing.

File: indexer/secure_index_builder.go (hunk 11 above)
Comment (jxguan) [inline indexer/secure_index_builder.go]: Ah I was thinking about needing to know that for serialization, but I was being stupid.

Fixed

File: client/client.go [hunk 21]
```diff
@@ -0,0 +1,142 @@
...
+type Client struct {
+	directory    string                      // Directory for the client where all the files are stored
+	server        *server.Server              // The server that this client is connected to
+	indexer       *indexer.SecureIndexBuilder // The indexer for the client
+	lookupTable   map[string]string           // A map from document ids to actual filenames
+	reverseLookup map[string]string           // A map from actual filenames to document ids
+}
+
+// CreateClient instantiates a client connected to Server `s` with a
+// `clientNum`.
```
Comment (strib) [inline client/client.go:30]: Might be worth mentioning here that the client instance can't be saved and reused after a switch to another client, since there's nothing to force the re-reading of the lookup table.  A new client must be constructed on every client switch.

File: server/server.go (hunk 15 above)
Comment (jxguan) [inline server/server.go]: Fixed

File: logger/logger.go (hunk 14 above)
Comment (jxguan) [inline logger/logger.go]: Fixed

File: logger/logger.go (hunk 13 above)
Comment (jxguan) [inline logger/logger.go]: Fixed

File: logger/logger.go (hunk 12 above)
Comment (jxguan) [inline logger/logger.go]: Fixed

File: server/server.go (hunk 18 above)
Comment (jxguan) [inline server/server.go]: Fixed

File: server/server.go (hunk 16 above)
Comment (jxguan) [inline server/server.go:120]: Fixed

We have to use time.Millisecond, because it only takes integer units

**Comment by strib** [comment]
Very nice job overall, thanks!  Mostly nits inline.

In general I'd like to see errors propagated all the way to main.go, and then printed out there.

The latency calculations are a little off, since there's a round-trip-time for every block of the file (plus another one for the top indirect block on big files), not just one RTT for the entire file.  I guess in the actual implementation we can try to optimize by fetching the blocks all at once, reducing it to about two RTTs for big files.

Anyway, once my comments are addressed, feel free to merge it.

File: indexer/secure_index_builder.go (hunk 10 above)
Comment (jxguan) [inline indexer/secure_index_builder.go]: Fixed

File: server/server_test.go (hunk 19 above)
Comment (jxguan) [inline server/server_test.go:172]: Acknowledged

File: client/client.go (hunk 20 above)
Comment (jxguan) [inline client/client.go:50]: Acknowledged

File: client/client.go (hunk 21 above)
Comment (jxguan) [inline client/client.go:30]: Fixed

File: index/secure_index.go (hunk 5 above)
Comment (jxguan) [inline index/secure_index.go]: Acknowledged

Will fix that in a separate PR.
